File: algorithms/ruby/0443-string-compression.rb
# frozen_string_literal: true

# Problem: 443. String Compression
# URL: https://leetcode.com/problems/string-compression

# @param {Character[]} chars
# @return {Integer}
def compress(chars)
  i = 0
  len = 0
  count = 0

  while i < chars.length do
    count += 1

    if chars[i + 1] != chars[i]
      chars[len] = chars[i]
      len += 1

      if count > 1
        count.to_s.each_char do |char|
          chars[len] = char
          len += 1
        end
      end

      count = 0
    end

    i += 1
  end

  len
end

# **************** #
#       TEST       #
# **************** #

require "test/unit"
class Test_compress < Test::Unit::TestCase
  def test_
    assert_equal 6, compress(["a", "a", "b", "b", "c", "c", "c"])
    assert_equal 1, compress(["a"])
    assert_equal 4, compress(["a", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b"])
  end
end

File: algorithms/python3/0443-string-compression.py
# Problem: 443. String Compression
# URL: https://leetcode.com/problems/string-compression

class Solution:
  def compress(self, chars: list[str]) -> int:
    n = len(chars)
    
    if n == 1:
      return 1

    i = 0
    j = 0
    count = 0

    while i < n:
      count += 1

      if i == n - 1 or chars[i + 1] != chars[i]:
        chars[j] = chars[i]
        j += 1

        if count > 1:
          for char in str(count):
            chars[j] = char
            j += 1

        count = 0

      i += 1

    return j

# ********************#
#       TEST         #
# ********************#

import unittest

class TestStringMethods(unittest.TestCase):
  def test_addBinary(self):
    self.assertEqual(Solution.compress(self, ["a", "a", "b", "b", "c", "c", "c"]), 6)
    self.assertEqual(Solution.compress(self, ["a"]), 1)
    self.assertEqual(Solution.compress(self, ["a", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b"]), 4)

if __name__ == '__main__':
    unittest.main()